iac-extensions/arm/src/main/java/org/sonar/iac/arm/visitors/ArmSymbolVisitor.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * SonarQube IaC Plugin
+ * Copyright (C) 2021-2024 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.iac.arm.visitors;
+
+import java.util.EnumSet;
+import org.sonar.iac.arm.symbols.SymbolTable;
+import org.sonar.iac.arm.symbols.Usage;
+import org.sonar.iac.arm.tree.api.ArmTree;
+import org.sonar.iac.arm.tree.api.File;
+import org.sonar.iac.arm.tree.api.Identifier;
+import org.sonar.iac.arm.tree.api.VariableDeclaration;
+import org.sonar.iac.common.extension.visitors.InputFileContext;
+import org.sonar.iac.common.extension.visitors.TreeVisitor;
+
+/**
+ * Class used to visit a ArmTree and build symbols and their usages for variables.
+ * Those Symbol/Usage can later be used in checks to resolve Argument(really?) or to report issue variable flow.
+ */
+
+public class ArmSymbolVisitor extends TreeVisitor<InputFileContext> {
+  private static final EnumSet<ArmTree.Kind> PARENT_KIND_INDICATING_NO_USAGE = EnumSet.of(
+    ArmTree.Kind.VARIABLE_DECLARATION,
+    ArmTree.Kind.OUTPUT_DECLARATION,
+    ArmTree.Kind.PROPERTY);
+  private final SymbolTable symbolTable = new SymbolTable();
```

Comment:
From my understanding, this will give all analyzed files the same `SymbolTable` instance.
I don't think this is expected, I would move this new `SymbolTable()` into `visitFile()`.
Also, I would add a test to show that two files analyzed get different `SymbolTable` and that those symbols does not reference each other.

Reply:
Good catch!